clippy_lints/src/eta_reduction.rs

```diff
@@ -170,15 +171,25 @@ impl<'tcx> LateLintPass<'tcx> for EtaReduction {
                 {
                     span_lint_and_then(cx, REDUNDANT_CLOSURE, expr.span, "redundant closure", |diag| {
                         if let Some(mut snippet) = snippet_opt(cx, callee.span) {
-                            if let Ok((ClosureKind::FnMut, _)) = cx.tcx.infer_ctxt().build().type_implements_fn_trait(
-                                cx.param_env,
-                                Binder::bind_with_vars(callee_ty_adjusted, List::empty()),
-                                ty::PredicatePolarity::Positive,
-                            ) && path_to_local(callee).map_or(false, |l| {
+                            if path_to_local(callee).map_or(false, |l| {
+                                // FIXME: Do we really need this `local_used_in` check?
+                                // Isn't it checking something like... `callee(callee)`?
+                                // If somehow this check is needed, add some test for it,
+                                // 'cuz currently nothing changes after deleting this check.
                                 local_used_in(cx, l, args) || local_used_after_expr(cx, l, expr)
                             }) {
```

Comment:
confused, but too scared to remove it

Reply:
Yeah, I'm also not entirely sure what that is for. Seems like that additional check was introduced in #8685, but that was a rather large refactor. Before this, there was only the local_used_after_expr check. Unfortunate that there's no test for this.

Reply:
Oh, it might mean situation like this?

```rust
let fun = |x: i32| -> i32 {
    x
};

let _ = fun(fun(0));
```

but I'm not sure, I'll test it.

Reply:
I don't *think* so. There is a check somewhere in the if chain that makes sure that the fn arguments are only paths to locals. So in your example that wouldn't happen because `fun(0)` is more than just a path.
https://github.com/rust-lang/rust-clippy/blob/03654badfd7221bd637daa7a214a0d73db176299/clippy_lints/src/eta_reduction.rs#L128-L131

https://github.com/rust-lang/rust-clippy/blob/03654badfd7221bd637daa7a214a0d73db176299/clippy_lints/src/eta_reduction.rs#L230-L234

Reply:
right... then I'm afraid that I can't think of any examples that pass the compilation check😂